source/serial_api.c

```diff
@@ -470,4 +500,238 @@ void serial_break_clear(serial_t *obj)
+    } else {
+        handle->State = HAL_UART_STATE_BUSY_RX;
+    }
+
+    __HAL_UART_CLEAR_PEFLAG(handle);
+    handle->Instance->CR1 |= USART_CR1_RXNEIE | USART_CR1_PEIE;
+    handle->Instance->CR3 |= USART_CR3_EIE;
+
+    DEBUG_PRINTF("UART%u: Rx: 0=(%u, %u, %u) %x\n", obj->serial.module+1, rx_length, rx_width, char_match, HAL_UART_GetState(handle));
+}
+
+int serial_irq_handler_asynch(serial_t *obj)
+{
+    UART_HandleTypeDef *handle = &UartHandle[obj->serial.module];
+
+    int status = handle->Instance->SR;
+    int data = handle->Instance->DR;
+    int event = 0;
+
+    if (status & USART_SR_PE) {
+        event |= SERIAL_EVENT_RX_PARITY_ERROR;
+    }
+    if (status & (USART_SR_NE | USART_SR_FE)) {
+        event |= SERIAL_EVENT_RX_FRAMING_ERROR;
+    }
+    if (status & USART_SR_ORE) {
+        event |= SERIAL_EVENT_RX_OVERRUN_ERROR;
+    }
+
+    if ((status & USART_SR_TC) && (handle->State & 0x10) && !handle->TxXferCount) {
+        // transmission is finally complete
+        handle->Instance->CR1 &= ~USART_CR1_TCIE;
+        // set event tx complete
+        event |= SERIAL_EVENT_TX_COMPLETE;
+        // update handle state
+        if(handle->State == HAL_UART_STATE_BUSY_TX_RX) {
+            handle->State = HAL_UART_STATE_BUSY_RX;
+        } else {
+            handle->State = HAL_UART_STATE_READY;
+        }
+    }
+    else if ((status & USART_SR_TXE) && handle->TxXferCount) {
+        // chose either the tx reg empty or if last byte wait directly for tx complete
+        if (--handle->TxXferCount == 0) {
+            handle->Instance->CR1 &= ~USART_CR1_TXEIE;
+            handle->Instance->CR1 |= USART_CR1_TCIE;
+        }
+        // copy new data into transmit register
+        handle->Instance->DR = (uint8_t)*handle->pTxBuffPtr++;
+        obj->tx_buff.pos++;
+    }
+
+    if ((status & USART_SR_RXNE) && handle->RxXferCount) {
+        // something arrived in the receive buffer
+        // copy into buffer
+        *handle->pRxBuffPtr++ = (uint8_t)data;
+        obj->rx_buff.pos++;
+        // check for character match only if enabled though!
+        if ((obj->serial.char_match != SERIAL_RESERVED_CHAR_MATCH) && ((uint8_t)data == obj->serial.char_match)) {
+            event |= SERIAL_EVENT_RX_CHARACTER_MATCH;
+        }
+        if (--handle->RxXferCount == 0) {
+            // last receive byte, disable all rx interrupts
+            handle->Instance->CR1 &= ~(USART_CR1_RXNEIE | USART_CR1_PEIE);
+            handle->Instance->CR3 &= ~USART_CR3_EIE;
+            // set event rx complete
+            event |= SERIAL_EVENT_RX_COMPLETE;
+            // update handle state
+            if(handle->State == HAL_UART_STATE_BUSY_TX_RX) {
+                handle->State = HAL_UART_STATE_BUSY_TX;
+            } else {
+                handle->State = HAL_UART_STATE_READY;
+            }
+        }
+    }
+
+    return (event & obj->serial.event);
+}
+
+void serial_tx_abort_asynch(serial_t *obj)
```

Comment:
Is it just me or did I really accidentally implement `serial_tx_abort_asynch` as `serial_rx_abort_asynch` and vice versa????
@0xc0170 @bogdanm @bremoran